vispy/visuals/filters/clipping_planes.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+
+from functools import lru_cache
+
+import numpy as np
+
+from ..shaders import Function, Varying
+from .base_filter import Filter
+
+
+class PlanesClipper(Filter):
+    """Clips visual output based on arbitrary clipping planes."""
+
+    VERT_CODE = """
+    void clip() {
+        // Transform back to visual coordinates and clip based on that
+        $v_is_shown = $clip_with_planes($itransform(gl_Position).xyz);
+    }
+    """
+
+    FRAG_CODE = """
+    void clip() {
+        if ($v_is_shown < 0.)
+            discard;
+    }
+    """
+
+    def __init__(self, clipping_planes=None, coord_system='scene'):
+        tr = ['visual', 'scene', 'document', 'canvas', 'framebuffer', 'render']
+        if coord_system not in tr:
+            raise ValueError(f'Invalid coordinate system {coord_system}. Must be one of {tr}.')
+        self._coord_system = coord_system
+
+        super().__init__(
+            vcode=Function(self.VERT_CODE), vhook='post', vpos=1,
+            fcode=Function(self.FRAG_CODE), fhook='pre', fpos=1,
+        )
+
+        v_is_shown = Varying('v_is_shown', 'float')
+        self.vshader['v_is_shown'] = v_is_shown
+        self.fshader['v_is_shown'] = v_is_shown
+
+        self.clipping_planes = clipping_planes
+
+    @property
+    def coord_system(self):
+        # unsettable cause we can't update the transform after being attached
+        return self._coord_system
+
+    def _attach(self, visual):
+        super()._attach(visual)
+        self.vshader['itransform'] = visual.get_transform('render', self._coord_system)
+
+    @staticmethod
+    @lru_cache(maxsize=10)
+    def _build_clipping_planes_func(n_planes):
+        """Build the code snippet used to clip the volume based on self.clipping_planes."""
+        func_template = '''
+            float clip_planes(vec3 loc) {{
+                float is_shown = 3.4e38; // max float
+                {clips};
+                return is_shown;
+            }}
+        '''
+        # the vertex is considered clipped if on the "negative" side of the plane
+        clip_template = '''
+            vec3 relative_vec{idx} = loc - $clipping_plane_pos{idx};
+            float is_shown{idx} = dot(relative_vec{idx}, $clipping_plane_norm{idx});
+            is_shown = min(is_shown{idx}, is_shown);
+            '''
+        all_clips = []
+        for idx in range(n_planes):
+            all_clips.append(clip_template.format(idx=idx))
+        formatted_code = func_template.format(clips=''.join(all_clips))
+        return Function(formatted_code)
+
+    @property
+    def clipping_planes(self):
+        """Get the set of planes used to clip the mesh.
+        Each plane is defined by a position and a normal vector (magnitude is irrelevant). Shape: (n_planes, 2, 3)
+        """
+        return self._clipping_planes[:, :, ::-1]
```

Comment:
This `::-1` was to make sure that the coordinates would be consistent on user side between clipping planes and volume. ~volume data in vispy is currently in `zyx` order, while stuff like points is in `xyz`.~  On closer inspection, this inconsistency is worse than I thought... The XYZ axis visual is doing some weird stuff, and this thing *should not* work with markers (since they are xyz), and yet it does. Something is fishy :/

Reply:
Ok, played aroudn with it a bit, and the axes transform callback is just wrong; good sign, lines are at least consistent with points.

Reply:
Ok, solved the issue and removed all the reversing. My confusion was ultimately caused by the fact that I was using the volume's data shape to create center for the clipping.

In short: volume `data` assumes `zyx` coordinates as input, thus reverses everything on the setter. This means that after that, everything should be assumed in `xyz`, even for volume clipping.